homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -1,207 +1,159 @@
 """Train information for departures and delays, provided by Trafikverket."""
 from __future__ import annotations
 
-from datetime import date, datetime, time, timedelta
-import logging
-from typing import TYPE_CHECKING, Any
-
-from pytrafikverket import TrafikverketTrain
-from pytrafikverket.exceptions import (
-    MultipleTrainAnnouncementFound,
-    NoTrainAnnouncementFound,
+from collections.abc import Callable
+from dataclasses import dataclass
+from datetime import datetime
+
+from homeassistant.components.sensor import (
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
 )
-from pytrafikverket.trafikverket_train import StationInfo, TrainStop
-
-from homeassistant.components.sensor import SensorDeviceClass, SensorEntity
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import CONF_NAME, CONF_WEEKDAY, WEEKDAYS
-from homeassistant.core import HomeAssistant
+from homeassistant.const import CONF_NAME
+from homeassistant.core import HomeAssistant, callback
 from homeassistant.helpers.device_registry import DeviceEntryType
 from homeassistant.helpers.entity import DeviceInfo
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
-from homeassistant.util import dt as dt_util
-
-from .const import CONF_FROM, CONF_TIME, CONF_TO, DOMAIN
-from .util import create_unique_id
-
-_LOGGER = logging.getLogger(__name__)
-
-ATTR_DEPARTURE_STATE = "departure_state"
-ATTR_CANCELED = "canceled"
-ATTR_DELAY_TIME = "number_of_minutes_delayed"
-ATTR_PLANNED_TIME = "planned_time"
-ATTR_ESTIMATED_TIME = "estimated_time"
-ATTR_ACTUAL_TIME = "actual_time"
-ATTR_OTHER_INFORMATION = "other_information"
-ATTR_DEVIATIONS = "deviations"
-
-ICON = "mdi:train"
-SCAN_INTERVAL = timedelta(minutes=5)
+from homeassistant.helpers.typing import StateType
+from homeassistant.helpers.update_coordinator import CoordinatorEntity
+
+from .const import ATTRIBUTION, DOMAIN
+from .coordinator import TVDataUpdateCoordinator
+
+
+@dataclass
+class TrafikverketRequiredKeysMixin:
+    """Mixin for required keys."""
+
+    value_fn: Callable[[dict[str, StateType | datetime]], StateType | datetime]
+
+
+@dataclass
+class TrafikverketSensorEntityDescription(
+    SensorEntityDescription, TrafikverketRequiredKeysMixin
+):
+    """Describes Trafikverket sensor entity."""
+
+
+SENSOR_TYPES: tuple[TrafikverketSensorEntityDescription, ...] = (
+    TrafikverketSensorEntityDescription(
+        key="departure_time",
+        translation_key="departure_time",
+        icon="mdi:clock",
+        device_class=SensorDeviceClass.TIMESTAMP,
+        value_fn=lambda data: data["departure_time"],
+    ),
+    TrafikverketSensorEntityDescription(
+        key="departure_state",
+        translation_key="departure_state",
+        icon="mdi:clock",
+        value_fn=lambda data: data["departure_state"],
+        device_class=SensorDeviceClass.ENUM,
+        options=["on_time", "delayed", "canceled"],
```

Comment:
What english do we use at HA? If it's british we should do cancelled

Reply:
It should be american english but it doesn't matter in the options, that only matters in the translations in that case.